ProcessOptimizer/samplers/doe/optimal_design.py

```diff
@@ -0,0 +1,697 @@
+            Builds a design for this order model.
+            Mutually exclusive with the **model** parameter.
+        * **model** (`patsy formula <https://patsy.readthedocs.io>`_) --
+            Builds a design for this model formula.
+            Mutually exclusive with the **order** parameter.
+        * **n_exp** (`integer`) --
+            The number of runs to use in the design. This must be equal
+            to or greater than the rank of the model.
+        * **space** (:class:`Space <ProcessOptimizer.space.Space>`) --
+            The space of the factors. This is used to be more consistent
+            when using categorical variables.
+        * **res** (`integer`) --
+            The resolution of the design. This is the sampling resolution used
+            when sampling the design space. The higher the resolution, the
+            more accurate the design will be. The default is 11.
+        * **seed** (`integer`) --
+            The seed to use for the random number generator. This is useful if
+            you want to reproduce the same design multiple times.
+
+    _______________________________________________________
+
+    This function is adapted from https://github.com/statease/dexpy
+    Copyright 2016 Stat-Ease, Inc.
+    License: Apache License, Version 2.0
+    License link: https://github.com/statease/dexpy/blob/master/LICENSE
+    """
+
+    model = kwargs.get("model", None)
+    include_powers = kwargs.get("include_powers", True)
+    space = kwargs.get("space", None)
+    res = kwargs.get("res", 11)
+
+    if model is None:
+        order = kwargs.get("order", 2)
+        model = make_model(factor_names, order, include_powers=include_powers)
+
+    n_exp = kwargs.get("n_exp", 0)
+    nseed = kwargs.get("seed", None)
+
+    # first generate a valid starting design
+    (design, X) = bootstrap(factor_names, model, n_exp, seed=nseed)
+
+    # The numbers in the design representing categorical factors should
+    # initially be exactly at one of the levels
+
+    if space is not None:
+        design = initial_values_cat_vars(design, space)
+
+    # Make sure that design point values are rounded to the resolution
+    design = round_design_point_values(design, res)
+
+    # Enable conversion between design points and X matrix
+    code = conversion_design_Xmatrix(X)
+
+    opt_design = optimize_design(
+        X, design, factor_names, code, res=res, space=space
+    )
+
+    return opt_design
+
+
+def model_order_and_include_powers(design_type):
+    """
+    Determine the order and include_powers of the model based on the design
+    type
+
+    :param design_type: The design_type of design to create.
+    :type design_type: str
+    :options: 'linear', 'screening', 'response', 'optimization'
+
+    :return: The order and include_powers of the model
+    :rtype: int, bool
+    """
+
+    # Specify options for the design types
+    design_types = ["linear", "screening", "response", "optimization"]
```

Comment:
```suggestion
    design_types = {"linear": (1, None), "screening": (2, False), "response": (2, True), "optimization": (3, True), None: (None, None)}
    # Key is design type, value[0] is order, value[1] is include powers
    if design_type not in design_types:
        raise ValueError(f"design_type must be one of {design_types.keys()}")
   order = design_types[design_type][0]
   include_powers = desing_types[design_type][1]
```
I don't like un-related lists that are secretly related (OK, that was too harsh). But it might just be a pet peeve.

Reply:
Took and adapted suggestion